File: test/cpp/test_elements.cpp
#include "test_utils.hpp"

#include "reactions/all.hpp"

using namespace reactions;

int main() {

  test::collector string_element_coll("string_element");
  string_element_coll.add_test_function("test string element",
                                        []() -> test::errors {
                                          test::errors errors;
                                          string_element el{"A"};
                                          el = "B";
                                          return errors;
                                        });

  test::collector pdg_element_coll("pdg_element");
  pdg_element_coll.add_test_function("test PDG element", []() -> test::errors {
    test::errors errors;

    try {

      auto &db = pdg_database::instance();

      if (db("pi+").pdg_id() != db(+211).pdg_id())
        errors.push_back(
            "Accessors by string and PDG ID do not retrieve the same elements");

      pdg_element::base_type ta = {
          "a", 0, 0, reactions::missing, reactions::missing, false};
      pdg_element a = std::move(ta);

      if ( !a.has<pdg::name>() || !a.has<pdg::pdg_id>() || a.has<pdg::mass>() )
	errors.push_back("Errors checking the presence of values");

      auto const& name = a.get<pdg::name>();

      if ( name != a.name() )
	errors.push_back("Name accessed with the \"name\" member function and with \"get<pdg::name>\" do not coincide");

      a.get<pdg::mass>(); // simply try to get it

      pdg_element::base_type tb = {
          "b", 0, 0, reactions::missing, reactions::missing, false};
      pdg_element b{std::move(tb)};

      pdg_element c1 = {"c1", 0, 0, reactions::missing, reactions::missing,
                        false};
      pdg_element c2 = {
          "c2", 0, 0, reactions::fill{0., 0., 0.}, reactions::missing, false};
      pdg_element c3 = {
          "c3", 0, 0, reactions::fill{0., 0., 0.}, reactions::fill{0., 0., 0.},
          false};

      if (c1.has_mass() || c1.has_width() || !c2.has_mass() || c2.has_width() ||
          !c3.has_mass() || !c3.has_width())
        errors.push_back("Problems found building custom PDG elements");

      auto check_latex_name = [&db, &errors](std::string const &name,
                                             std::string const &reference) {
        auto const ln = db(name).latex_name();
        if (ln != reference)
          errors.push_back("Wrong LaTeX name for particle \"" + name +
                           "\": \"" + ln + "\" (reference: \"" + reference +
                           "\")");
      };

      try {
        check_latex_name("K(S)0", "K_{S}^{0}");
        check_latex_name("K+", "K^{+}");
        check_latex_name("pi+", "\\pi^{+}");
        check_latex_name("pi-", "\\pi^{-}");
        check_latex_name("Lambda", "\\Lambda");
        check_latex_name("eta'(958)", "\\eta^{'}(958)");
        check_latex_name("a(0)(980)0", "a_{0}(980)^{0}");
        check_latex_name("f(2)'(1525)", "f_{2}^{'}(1525)");
        check_latex_name("K(2)*(1430)~0", "\\bar{K}_{2}^{*}(1430)^{0}");
        check_latex_name("D(s2)*(2573)+", "D_{s2}^{*}(2573)^{+}");
        check_latex_name("Xi(c)'+", "\\Xi_{c}^{'+}");
        check_latex_name("Delta(1950)~-", "\\bar{\\Delta}(1950)^{-}");
      } catch (reactions::internal_error &) {
        errors.push_back("Internal error detected processing LaTeX names");
      }

      try {
        for (auto e : reactions::pdg_database::instance().all_elements())
          e.latex_name();
      } catch (reactions::internal_error &) {
        errors.push_back("Unable to compute LaTeX names for some elements");
      }
    }
    REACTIONS_TEST_UTILS_CATCH_EXCEPTIONS(errors);

    return errors;
  });

  auto str_status = !string_element_coll.run();
  auto pdg_status = !pdg_element_coll.run();

  return str_status || pdg_status;
}

File: test/python/test_elements.py
"""
Test element classes
"""
import pytest
import reactions


def test_nubase_element():

    # create an element from the database
    el = reactions.nubase_element('1H')
    assert reactions.node_type(el) == 'element'
    assert el.name == '1H'
    assert el.nubase_id == 1001000

    el = reactions.nubase_element(1001000)
    assert reactions.node_type(el) == 'element'

    # check that empty values in C++ correspond to None in python
    assert reactions.nubase_element('1H').half_life is None
    assert reactions.nubase_element('1H').half_life_error is None
    assert reactions.nubase_element('1H').half_life_from_systematics is None

    # create a custom element
    n0 = reactions.nubase_element(
        '999Un', 999999000, 999, 999, (0., 0., False), False, None, True)
    n1 = reactions.nubase_element(name='999Un', nubase_id=999999000, atomic_number=999, mass_number=999, mass_excess_and_error_with_tag=(
        0., 0., False), half_life_and_error_with_tag=None, is_stable=True, is_ground_state=True)

    assert n0 == n1

    assert str(n0) == n0.__repr__()

    # all these constructors must fail
    with pytest.raises(TypeError):
        reactions.nubase_element(0.0)

    with pytest.raises(RuntimeError):
        reactions.nubase_element('gamma', 1, 22)

    with pytest.raises(RuntimeError):
        reactions.nubase_element('gamma', 1, 22, nubase_id=22)

    for e in reactions.nubase_database.all_elements():
        # we must be able to compute all the LaTeX names
        e.latex_name

    # test LaTeX names
    assert reactions.nubase_element("1H").latex_name == "\\ce{^{1}H}"
    assert reactions.nubase_element("1n").latex_name == "\\ce{^{1}n}"
    assert reactions.nubase_element("7Li(i)").latex_name == "\\ce{^{7}Li^{i}}"

    # errors accessing elements
    with pytest.raises(reactions.LookupError):
        reactions.nubase_element('H')


def test_pdg_element():

    # create an element from the database
    el = reactions.pdg_element('pi+')
    assert reactions.node_type(el) == 'element'
    assert el.name == 'pi+'
    assert el.pdg_id == 211

    el = reactions.pdg_element(211)
    assert reactions.node_type(el) == 'element'

    # check that empty values in C++ correspond to None in python
    assert reactions.pdg_element('H0').width is None
    assert reactions.pdg_element('H0').width_error_lower is None
    assert reactions.pdg_element('H0').width_error_upper is None
    assert reactions.pdg_element('H0').width_error is None

    # create a custom element
    reactions.pdg_element('gamma', 1, 0, (0., 0., 0.), (0., 1.e+16, 0.), True)
    reactions.pdg_element(name='gamma', pdg_id=1, three_charge=0, mass_and_errors=(
        0., 0., 0.), width_and_errors=(0., 0., 0.), is_self_cc=True)
    reactions.pdg_element(name='gamma', pdg_id=1, three_charge=0,
                          mass_and_errors=None, width_and_errors=(0., 0., 0.), is_self_cc=True)
    g0 = reactions.pdg_element(name='gamma', pdg_id=1, three_charge=0, mass_and_errors=(
        0., 0., 0.), width_and_errors=None, is_self_cc=False)

    assert str(g0) == g0.__repr__()

    g1 = reactions.pdg_element(name='gamma', pdg_id=1, three_charge=0,
                               mass_and_errors=None, width_and_errors=None, is_self_cc=True)

    assert str(g1) == 'reactions.pdg_element(name="gamma", pdg_id=1, three_charge=0, mass_and_errors=None, width_and_errors=None, is_self_cc=True)'
    assert str(g1) == g1.__repr__()

    # all these constructors must fail
    with pytest.raises(TypeError):
        reactions.pdg_element(0.0)

    with pytest.raises(RuntimeError):
        reactions.pdg_element('gamma', 1, 22)

    with pytest.raises(RuntimeError):
        reactions.pdg_element('gamma', 1, 22, pdg_id=22)

    # test the comparison operators
    assert reactions.pdg_element('pi+') == reactions.pdg_element(+211)
    assert reactions.pdg_element('pi+') != reactions.pdg_element('pi-')
    assert reactions.pdg_element(+211) != reactions.pdg_element(-211)

    # errors accessing elements
    with pytest.raises(reactions.LookupError):
        reactions.pdg_element('mu')

    # test LaTeX names
    assert reactions.pdg_element("K(S)0").latex_name == "K_{S}^{0}"
    assert reactions.pdg_element("K+").latex_name == "K^{+}"
    assert reactions.pdg_element("pi+").latex_name == "\\pi^{+}"
    assert reactions.pdg_element("pi-").latex_name == "\\pi^{-}"
    assert reactions.pdg_element("Lambda").latex_name == "\\Lambda"
    assert reactions.pdg_element("eta'(958)").latex_name == "\\eta^{'}(958)"
    assert reactions.pdg_element("a(0)(980)0").latex_name == "a_{0}(980)^{0}"
    assert reactions.pdg_element("f(2)'(1525)").latex_name == "f_{2}^{'}(1525)"
    assert reactions.pdg_element("Xi(c)'+").latex_name == "\\Xi_{c}^{'+}"
    assert reactions.pdg_element(
        "Delta(1950)~-").latex_name == "\\bar{\\Delta}(1950)^{-}"
    assert reactions.pdg_element(
        "K(2)*(1430)~0").latex_name == "\\bar{K}_{2}^{*}(1430)^{0}"
    assert reactions.pdg_element(
        "D(s2)*(2573)+").latex_name == "D_{s2}^{*}(2573)^{+}"

    for e in reactions.pdg_database.all_elements():
        # we must be able to compute all the LaTeX names
        e.latex_name


def test_string_element():

    el = reactions.string_element('custom')
    assert el.name == 'custom'
    assert reactions.node_type(el) == 'element'
    assert reactions.string_element(
        'custom') == reactions.string_element('custom')
    assert reactions.string_element(
        'custom') != reactions.string_element('other')
    assert str(el) == 'reactions.string_element(name="custom")'
